INT-3831: Fix MessagingGatewaySupport MBean Export

garyrussell opened the pull request:
JIRA: https://jira.spring.io/browse/INT-3831
- Add `@IntegrationManagedResource`
- Suppress exporting as an endpoint

garyrussell committed:
INT-3831: Fix MessagingGatewaySupport MBean Export

JIRA: https://jira.spring.io/browse/INT-3831

- Add `@IntegrationManagedResource`
- Suppress exporting as an endpoint

garyrussell commented:
repushed

artembilan committed:
Fix `outputChannel` issue for anonymous `MessagingGatewaySupport`

artembilan commented:
Good, but thanks to check against DSL we have an issue on the next step:
```
Caused by: java.lang.IllegalArgumentException: Could not find field [outputChannel] on target [org.springframework.integration.jmx.config.MBeanRegistrationTests.MyMessagingGateway#0]
	at org.springframework.integration.monitor.IntegrationMBeanExporter.getField(IntegrationMBeanExporter.java:1067)
```

to reproduce it just remove the `id` from your `MyMessagingGateway` bean definition in test-case.

So, here is a polishing from my perspective: https://github.com/artembilan/spring-integration/commit/148ca5c60a8d319ef34d3b030b9d567bf2015fa5.

artembilan commented:
`BUILD SUCCESSFUL`
On my DSL PR. 
BTW the issue there I had in the `JmsTests`, where `IntegrationMBeanExporter` is initiated by the `IntegrationAutoConfiguration` and the `JmsInboundGateway` is created by the
```
IntegrationFlows
	.from((MessagingGateways g) -> g
		.jms(this.jmsConnectionFactory)
		.destination("jmsPipelineTest"))
```
as an anonymous endpoint.

garyrussell committed:
INT-3831: Pure JavaConfig Test

garyrussell commented:
Thanks; I'll take a look at your changes. In the meantime; do you have any idea why this test (pushed) fails with gradle? It works fine in the IDE but in gradle it seems to create a new MBeanServer and causes other tests to fail.

garyrussell commented:
Good catch (output channel). However, don't we need to handle the fact that the output/request channel might be lazy (`...Name`) ??

garyrussell committed:
INT-3831: Polishing

Check output channel name for anonymous message source metrics.

Don't use getter on MGS to avoid early channel resolution.

garyrussell commented:
Pushed - still working on MBeanServer test failures.

artembilan commented:
>  However, don't we need to handle the fact that the output/request channel might be lazy (...Name) ??

That's why I made `MessagingGatewaySupport.getRequestChannel()` as `public`.
To call it here is valid because of `afterSingletonsInstantiated()` context.

I'll take a look to your `MBeanServer` issue after the breakfast

garyrussell commented:
But what about the other sources? (non-gateway).

artembilan commented:
Well, let's take a look what we have for the `AbstractEndpoint` and `MessageSourceMetrics` impls around:
* `MessagingGatewaySupport`
* `SourcePollingChannelAdapter`

That's all. Other `AbstractEndpoint` cases are out of scope of this `enhanceSourceMonitor()`.
Or do you see something else here on the matter?..

BTW:
```
outputChannel = getField(target, "outputChannel");
```
doesn't work for us. Because of:
```
Field field = ReflectionUtils.findField(target.getClass(), name);
		if (field == null) {
			throw new IllegalArgumentException("Could not find field [" + name + "] on target [" + target + "]");
		}
```

artembilan commented:
BTW. I test everything in the DSL with JMX, so, looks like `MessagingGatewaySupport` is only an issue.

artembilan commented:
> doesn't work for us. Because of:

Doh... We extract there exactly `field` not a value. Never mind

artembilan commented on spring-integration-jmx/src/test/java/org/springframework/integration/monitor/MessagingGatewayMetricsTests.java:
```diff
@@ -0,0 +1,119 @@
+
+	@Test
+	public void testHandlerMBeanRegistration() throws Exception {
+		System.out.println(server);
```
`SOUT`.
I guess you will remove it during the next push

garyrussell committed:
INT-3831: Fix Test - Avoid Second MBeanServer

Reference the `MBSFB` in the context rather than create a new `MBeanServer`.

artembilan commented:
With your `fb.setLocateExistingServerIfPossible(true);` I have even issue from IDE just only on test in the `org.springframework.integration.monitor` package:
```
com.sun.jmx.mbeanserver.JmxMBeanServer@4550bb58

java.lang.AssertionError: 
Expected :1
Actual   :0
```
And that's because `IntegrationMBeanExportConfiguration` don't use `server()` bean, but that one internal:
```
else {
	exporter.setServer(MBeanServerFactory.createMBeanServer());
}
```
From other side that  `server()` is used some other non-stopped server instance from other tests. Not that one internal. 
So, looks like an issue from other tests not this new one.

artembilan commented:
I think we aren't correct in the `IntegrationMBeanExportConfiguration` and should follow with the same logic in its legacy `MBeanExportConfiguration`:
```
else {
			MBeanExportConfiguration.SpecificPlatform specificPlatform = MBeanExportConfiguration.SpecificPlatform.get();
			if (specificPlatform != null) {
				exporter.setServer(specificPlatform.getMBeanServer());
			}
		}
```
after this fix everything starts work well. Even with and additional `@EnableMBeanExport` to the `MessagingGatewayMetricsTests`.

You should convince me regarding the reflection above and we are ready to merge this "last man standing"

garyrussell commented:
Right - I came to the same conclusion see https://jira.spring.io/browse/INT-3869

Pushed

garyrussell commented:
Since it's a behavior change; we can't do it in a point release.

garyrussell commented:
Simply adding a `server` attribute to the exporter solves the problem

artembilan commented on spring-integration-jmx/src/test/java/org/springframework/integration/monitor/MessagingGatewayMetricsTests.java:
```diff
@@ -0,0 +1,126 @@
+				@Override
+				protected void onInit() throws Exception {
+
+				}
```
I checked locally: this method override is really redundant.
Will fix on merge.
Sorry for the noise from my side, where you have copied this test.

artembilan commented on spring-integration-jmx/src/test/java/org/springframework/integration/monitor/MessagingGatewayMetricsTests.java:
```diff
@@ -0,0 +1,126 @@
+			return new NullChannel();
+		}
+
+		@Bean(name="org.springframework.integration.foo1")
```
With my pure vision this bean is almost `fool` :smile: 
Joke, never mind

artembilan commented:
Tests on DSL pass - merging...

artembilan commented:
Merged as https://github.com/spring-projects/spring-integration/commit/37aaa82272ddc310f5244ae66c5845fce52c6079
